<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="**A**tac **S**eq **P**ip**E**li**N**e"><meta name=author content=CCBR><link rel=icon href=assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.6.1, mkdocs-material-9.1.6"><title>ASPEN</title><link rel=stylesheet href=assets/stylesheets/main.ded33207.min.css><link rel=stylesheet href=assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#aspen class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <div data-md-color-scheme=default data-md-component=outdated hidden> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=. title=ASPEN class="md-header__button md-logo" aria-label=ASPEN data-md-component=logo> <img src=assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> ASPEN </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Home </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=red data-md-color-accent=red aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCBR/ASPEN title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/ASPEN </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=. class="md-tabs__link md-tabs__link--active"> Home </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=. title=ASPEN class="md-nav__button md-logo" aria-label=ASPEN data-md-component=logo> <img src=assets/icons/doc-book.svg alt=logo> </a> ASPEN </label> <div class=md-nav__source> <a href=https://github.com/CCBR/ASPEN title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> CCBR/ASPEN </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Home <span class="md-nav__icon md-icon"></span> </label> <a href=. class="md-nav__link md-nav__link--active"> Home </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#table-of-contents class=md-nav__link> Table of Contents </a> </li> <li class=md-nav__item> <a href=#1-outline class=md-nav__link> 1. Outline </a> </li> <li class=md-nav__item> <a href=#2-runtime-details class=md-nav__link> 2. Runtime details </a> <nav class=md-nav aria-label="2. Runtime details"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-load-module-on-biowulf class=md-nav__link> 2.1 Load module on BIOWULF </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#22-create-sample-manifest class=md-nav__link> 2.2 Create Sample Manifest </a> </li> <li class=md-nav__item> <a href=#23-run-pipeline class=md-nav__link> 2.3 Run Pipeline </a> </li> <li class=md-nav__item> <a href=#3-genomes class=md-nav__link> 3. Genomes </a> </li> <li class=md-nav__item> <a href=#4-disclaimer class=md-nav__link> 4. Disclaimer </a> </li> <li class=md-nav__item> <a href=#5-help class=md-nav__link> 5. Help </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <!-- the README is inserted here automatically--> <h1 id=aspen><strong>ASPEN</strong><a class=headerlink href=#aspen title="Permanent link">&para;</a></h1> <h3 id=table-of-contents>Table of Contents<a class=headerlink href=#table-of-contents title="Permanent link">&para;</a></h3> <ul> <li><a href=#aspen>ASPEN - <strong>A</strong>tac <strong>S</strong>eq <strong>P</strong>ip<strong>E</strong>li<strong>N</strong>e</a></li> <li><a href=#1-outline>1. Outline</a></li> <li><a href=#2-runtime-details>2. Runtime details</a><ul> <li><a href=#21-load-module-on-biowulf>2.1 Load Module On Biowulf</a></li> <li><a href=#22-create-sample-manifest>2.2 Create Sample Manifest</a></li> <li><a href=#23-run-pipeline>2.3 Run Pipeline</a></li> </ul> </li> <li><a href=#3-genomes>3. Genomes</a></li> <li><a href=#4-disclaimer>4. Disclaimer</a></li> <li><a href=#5-help>5. Help</a></li> </ul> <h3 id=1-outline>1. Outline<a class=headerlink href=#1-outline title="Permanent link">&para;</a></h3> <p>ASPEN or <strong>A</strong>tac <strong>S</strong>eq <strong>P</strong>ip<strong>E</strong>li<strong>N</strong>e is CCBR's pipeline to calls peaks for ATAC-Seq datasets. It currently accepts paired-end Illumina data and calls peak using <a href=https://doi.org/10.1186/gb-2008-9-9-r137>MACS2</a> and <a href=https://github.com/jsh58/Genrich>Genrich</a> peak callers. Below is a brief outline of the steps performed by the pipeline:</p> <ul> <li>Trim PE reads with <a href=https://doi.org/10.14806/ej.17.1.200>CutAdapt</a></li> <li>Remove reads aligning to known <a href=https://doi.org/10.1038/s41598-019-45839-z>blacklisted regions</a>, if provided</li> <li>Align reads provided genome using <a href=https://doi.org/10.1038%2Fnmeth.1923>bowtie2</a>. This step generates multiple output files:</li> <li><code>tagAlign.gz</code>, which is a BED6 format file mainly required for MACS2 peak calling</li> <li><code>dedup.bam</code>, deduplicated BAM format file which may be required for downstream processing (eg. <a href=https://github.com/CCBR/CCBR_tobias>TOBIAS</a>)</li> <li><code>qsorted.bam</code>, query sorted BAM file for Genrich peak calling</li> <li>Pre-peakcalling QC metrics:</li> <li><a href=https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ >FastQC</a> is run pre- and post-trimming</li> <li>Fragment length distribution is calculated using custom scripts</li> <li><a href=https://smithlabresearch.org/software/preseq/ >Preseq</a> is run to estimate library complexity</li> <li>Post-peakcalling QC metrics:</li> <li>TSS distributions are calculated for each replicate</li> <li>FRiP (Fraction of Reads in Peaks) is calculated for each replicate</li> <li>FRiPextra calculations are performed if <em>fripextra</em> config files are supplied<ul> <li>Fraction of reads in DHS regions</li> <li>Fraction of reads in promoter regions</li> <li>Fraction of reads in enhancer regions</li> </ul> </li> <li>Peak calling: Peaks (NarrowPeak format) are called using MACS2 and Genrich. If multiple replicates exist per sample, consensus peaks are called (BED format).</li> <li>Peak annotation: <a href=https://doi.org/10.1093/bioinformatics/btv145>ChIPseeker</a> is used to annotate peaks if the genome is hg38/hg19/mm10</li> <li>Motif enrichment: Motif Enrichment is calculated using <a href=http://homer.ucsd.edu/homer/ >HOMER</a> and <a href=https://meme-suite.org/meme/doc/ame.html>AME (MEME suite)</a></li> <li>Report: <a href=10.1093/bioinformatics/btw354>MultiQC</a> is used to generate a customized final HTML report</li> </ul> <h3 id=2-runtime-details>2. Runtime details<a class=headerlink href=#2-runtime-details title="Permanent link">&para;</a></h3> <h4 id=21-load-module-on-biowulf>2.1 Load module on BIOWULF<a class=headerlink href=#21-load-module-on-biowulf title="Permanent link">&para;</a></h4> <p>To clone the repo run:</p> <div class=highlight><pre><span></span><code>%<span class=w> </span>module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
</code></pre></div> <p>This will add <code>aspen</code> to your PATH environmental variable.</p> <blockquote> <p>NOTE: If not running on BIOWULF, please ensure that the required tools like snakemake, python and singularity are in PATH.</p> </blockquote> <h3 id=22-create-sample-manifest>2.2 Create Sample Manifest<a class=headerlink href=#22-create-sample-manifest title="Permanent link">&para;</a></h3> <p>Once the data is stored on biowulf, sample manifest TSV (<code>samples.tsv</code>) can be created to have the following columns:</p> <ol> <li> <p>replicateName</p> </li> <li> <p>sampleName: Multiple replicates can have the same sampleName</p> </li> <li> <p>path_to_R1_fastq: absolute path is preferred</p> </li> <li> <p>path_to_R2_fastq: <strong>Required!</strong> Pipeline currently <strong>only</strong> supports paired-end data</p> </li> </ol> <p>Note that:</p> <ul> <li>symlinks are created for R1 and R2 files from the sample manifest in the results folder. These symlinks have the filenames \&lt;replicateName&gt;.R1.fastq.gz and \&lt;replicateName&gt;.R2.fastq.gz, respectively. Thus, original filenames do not matter and original files do not need to be renamed.</li> <li><strong>replicateName</strong> is used as prefix for individual peak calls</li> <li><strong>sampleName</strong> is used as prefix for consensus peak calls</li> </ul> <blockquote> <p>NOTE: Optionally, if running differential ATAC please also provide <code>contrasts.tsv</code> in the output folder after running <code>init</code>. This is a simple tab-delimited text file with 2 columns (<em>Group1</em> and <em>Group2</em>) without any headers.</p> </blockquote> <h3 id=23-run-pipeline>2.3 Run Pipeline<a class=headerlink href=#23-run-pipeline title="Permanent link">&para;</a></h3> <p>To get more information about how to run the pipeline simply <code>cd</code> to the above <code>CCBR_ATACseq</code> folder and run</p> <div class=highlight><pre><span></span><code>%<span class=w> </span>aspen<span class=w> </span>--help
<span class=c1>##########################################################################################</span>

Welcome<span class=w> </span>to
____<span class=w> </span>____<span class=w> </span>___<span class=w>  </span>____<span class=w> </span>_<span class=w>  </span>_
<span class=p>|</span>__<span class=p>|</span><span class=w> </span><span class=o>[</span>__<span class=w>  </span><span class=p>|</span>__<span class=o>]</span><span class=w> </span><span class=p>|</span>___<span class=w> </span><span class=p>|</span><span class=se>\ </span><span class=p>|</span>
<span class=p>|</span><span class=w>  </span><span class=p>|</span><span class=w> </span>___<span class=o>]</span><span class=w> </span><span class=p>|</span><span class=w>    </span><span class=p>|</span>___<span class=w> </span><span class=p>|</span><span class=w> </span><span class=se>\|</span><span class=w>    </span>v1.0.0

A_TAC_S_eq<span class=w> </span>A_nalysis<span class=w> </span>P_ip_E_li_N_e

<span class=c1>##########################################################################################</span>

This<span class=w> </span>pipeline<span class=w> </span>was<span class=w> </span>built<span class=w> </span>by<span class=w> </span>CCBR<span class=w> </span><span class=o>(</span>https://bioinformatics.ccr.cancer.gov/ccbr<span class=o>)</span>
Please<span class=w> </span>contact<span class=w> </span>Vishal<span class=w> </span>Koparde<span class=w> </span><span class=k>for</span><span class=w> </span>comments/questions<span class=w> </span><span class=o>(</span>vishal.koparde@nih.gov<span class=o>)</span>

<span class=c1>##########################################################################################</span>

Here<span class=w> </span>is<span class=w> </span>a<span class=w> </span>list<span class=w> </span>of<span class=w> </span>genome<span class=w> </span>supported<span class=w> </span>by<span class=w> </span>aspen:

<span class=w>  </span>*<span class=w> </span>hg19<span class=w>          </span><span class=o>[</span>Human<span class=o>]</span>
<span class=w>  </span>*<span class=w> </span>hg38<span class=w>          </span><span class=o>[</span>Human<span class=o>]</span>
<span class=w>  </span>*<span class=w> </span>mm10<span class=w>          </span><span class=o>[</span>Mouse<span class=o>]</span>
<span class=w>  </span>*<span class=w> </span>mmul10<span class=w>        </span><span class=o>[</span>Macaca<span class=w> </span>mulatta<span class=o>(</span>Rhesus<span class=w> </span>monkey<span class=o>)</span><span class=w> </span>or<span class=w> </span>rheMac10<span class=o>]</span>
<span class=w>  </span>*<span class=w> </span>bosTau9<span class=w>       </span><span class=o>[</span>Bos<span class=w> </span>taurus<span class=o>(</span>cattle<span class=o>)]</span>

aspen<span class=w> </span>calls<span class=w> </span>peaks<span class=w> </span>using<span class=w> </span>the<span class=w> </span>following<span class=w> </span>tools:

<span class=w> </span>*<span class=w> </span>MACS2
<span class=w> </span>*<span class=w> </span>Genrich<span class=w>        </span><span class=o>[</span>RECOMMENDED<span class=w> </span>FOR<span class=w> </span>USE<span class=o>]</span>

USAGE:
<span class=w>  </span>bash<span class=w> </span>./aspen<span class=w> </span>-w/--workdir<span class=o>=</span>&lt;WORKDIR&gt;<span class=w> </span>-m/--runmode<span class=o>=</span>&lt;RUNMODE&gt;

Required<span class=w> </span>Arguments:
<span class=m>1</span>.<span class=w>  </span>WORKDIR<span class=w>     </span>:<span class=w> </span><span class=o>[</span>Type:<span class=w> </span>String<span class=o>]</span>:<span class=w> </span>Absolute<span class=w> </span>or<span class=w> </span>relative<span class=w> </span>path<span class=w> </span>to<span class=w> </span>the<span class=w> </span>output<span class=w> </span>folder<span class=w> </span>with<span class=w> </span>write<span class=w> </span>permissions.

<span class=m>2</span>.<span class=w>  </span>RUNMODE<span class=w>     </span>:<span class=w> </span><span class=o>[</span>Type:<span class=w> </span>String<span class=o>]</span><span class=w> </span>Valid<span class=w> </span>options:
<span class=w>    </span>*<span class=w> </span>init<span class=w>      </span>:<span class=w> </span>initialize<span class=w> </span>workdir
<span class=w>    </span>*<span class=w> </span>dryrun<span class=w>    </span>:<span class=w> </span>dry<span class=w> </span>run<span class=w> </span>snakemake<span class=w> </span>to<span class=w> </span>generate<span class=w> </span>DAG
<span class=w>    </span>*<span class=w> </span>run<span class=w>       </span>:<span class=w> </span>run<span class=w> </span>with<span class=w> </span>slurm
<span class=w>    </span>*<span class=w> </span>runlocal<span class=w>  </span>:<span class=w> </span>run<span class=w> </span>without<span class=w> </span>submitting<span class=w> </span>to<span class=w> </span>sbatch
<span class=w>    </span>ADVANCED<span class=w> </span>RUNMODES<span class=w> </span><span class=o>(</span>use<span class=w> </span>with<span class=w> </span>caution!!<span class=o>)</span>
<span class=w>    </span>*<span class=w> </span>unlock<span class=w>    </span>:<span class=w> </span>unlock<span class=w> </span>WORKDIR<span class=w> </span><span class=k>if</span><span class=w> </span>locked<span class=w> </span>by<span class=w> </span>snakemake<span class=w> </span>NEVER<span class=w> </span>UNLOCK<span class=w> </span>WORKDIR<span class=w> </span>WHERE<span class=w> </span>PIPELINE<span class=w> </span>IS<span class=w> </span>CURRENTLY<span class=w> </span>RUNNING!
<span class=w>    </span>*<span class=w> </span>reconfig<span class=w>  </span>:<span class=w> </span>recreate<span class=w> </span>config<span class=w> </span>file<span class=w> </span><span class=k>in</span><span class=w> </span>WORKDIR<span class=w> </span><span class=o>(</span>debugging<span class=w> </span>option<span class=o>)</span><span class=w> </span>EDITS<span class=w> </span>TO<span class=w> </span>config.yaml<span class=w> </span>WILL<span class=w> </span>BE<span class=w> </span>LOST!
<span class=w>    </span>*<span class=w> </span>reset<span class=w>     </span>:<span class=w> </span>DELETE<span class=w> </span>workdir<span class=w> </span>dir<span class=w> </span>and<span class=w> </span>re-init<span class=w> </span>it<span class=w> </span><span class=o>(</span>debugging<span class=w> </span>option<span class=o>)</span><span class=w> </span>EDITS<span class=w> </span>TO<span class=w> </span>ALL<span class=w> </span>FILES<span class=w> </span>IN<span class=w> </span>WORKDIR<span class=w> </span>WILL<span class=w> </span>BE<span class=w> </span>LOST!
<span class=w>    </span>*<span class=w> </span>printbinds:<span class=w> </span>print<span class=w> </span>singularity<span class=w> </span>binds<span class=w> </span><span class=o>(</span>paths<span class=o>)</span>
<span class=w>    </span>*<span class=w> </span><span class=nb>local</span><span class=w>     </span>:<span class=w> </span>same<span class=w> </span>as<span class=w> </span>runlocal

Optional<span class=w> </span>Arguments:

--help<span class=p>|</span>-h<span class=w>       </span>:<span class=w> </span>print<span class=w> </span>this<span class=w> </span><span class=nb>help</span>
--genome<span class=p>|</span>-g<span class=w>     </span>:<span class=w> </span>genome<span class=w> </span>eg.<span class=w> </span>hg38
--manifest<span class=p>|</span>-s<span class=w>   </span>:<span class=w> </span>absolute<span class=w> </span>path<span class=w> </span>to<span class=w> </span>samples.tsv.<span class=w> </span>This<span class=w> </span>will<span class=w> </span>be<span class=w> </span>copied<span class=w> </span>to<span class=w> </span>output<span class=w> </span>folder<span class=w>                    </span><span class=o>(</span>--runmode<span class=o>=</span>init<span class=w> </span>only<span class=o>)</span>
--useenvmod<span class=p>|</span>-e<span class=w>  </span>:<span class=w> </span>use<span class=w> </span><span class=s2>&quot;--use-enmodules&quot;</span><span class=w> </span>option<span class=w> </span><span class=k>while</span><span class=w> </span>running<span class=w> </span>Snakemake.<span class=w> </span>This<span class=w> </span>is<span class=w> </span><span class=k>for</span><span class=w> </span>using<span class=w> </span>modules<span class=w> </span>on<span class=w> </span>HPC<span class=w> </span>instead<span class=w> </span>of<span class=w> </span>containers<span class=o>(</span>default<span class=o>)</span>.
--singcache<span class=p>|</span>-c<span class=w>  </span>:<span class=w> </span>singularity<span class=w> </span>cache<span class=w> </span>directory.<span class=w> </span>Default<span class=w> </span>is<span class=w> </span><span class=sb>`</span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/.singularity<span class=sb>`</span><span class=w> </span><span class=k>if</span><span class=w> </span>available,<span class=w> </span>or<span class=w> </span>falls<span class=w> </span>back<span class=w> </span>to<span class=w> </span><span class=sb>`</span><span class=si>${</span><span class=nv>WORKDIR</span><span class=si>}</span>/.singularity<span class=sb>`</span>.


Example<span class=w> </span>commands:
<span class=w>  </span>bash<span class=w> </span>./aspen<span class=w> </span>-w<span class=o>=</span>/my/output/folder<span class=w> </span>-m<span class=o>=</span>init
<span class=w>  </span>bash<span class=w> </span>./aspen<span class=w> </span>-w<span class=o>=</span>/my/output/folder<span class=w> </span>-m<span class=o>=</span>dryrun
<span class=w>  </span>bash<span class=w> </span>./aspen<span class=w> </span>-w<span class=o>=</span>/my/output/folder<span class=w> </span>-m<span class=o>=</span>run
<span class=w>  </span>bash<span class=w> </span>./aspen<span class=w> </span>-w<span class=o>=</span>/my/output/folder<span class=w> </span>-m<span class=o>=</span>run<span class=w> </span>-c<span class=w> </span>/data/<span class=si>${</span><span class=nv>USER</span><span class=si>}</span>/.singularity

<span class=c1>##########################################################################################</span>

VersionInfo:
<span class=w>  </span>python<span class=w>          </span>:<span class=w> </span>python/3.10
<span class=w>  </span>snakemake<span class=w>       </span>:<span class=w> </span>snakemake
<span class=w>  </span>pipeline_home<span class=w>   </span>:<span class=w> </span>/gpfs/gsfs10/users/CCBR_Pipeliner/Pipelines/ASPEN/dev
<span class=w>  </span>git<span class=w> </span>commit/tag<span class=w>  </span>:<span class=w> </span>51cb3aee2142ce1226acca97e1e662d54c881a13<span class=w>    </span>v1.0.0-2-g51cb3ae
<span class=w>  </span>aspen_version<span class=w>   </span>:<span class=w> </span>v1.0.0

<span class=c1>##########################################################################################</span>
</code></pre></div> <p><code>aspen</code> is a <strong>Biowulf</strong> and <strong>FRCE</strong> specific wrapper script to the pipeline. Essentially, to run the pipeline the user has to follow 3 steps:</p> <ol> <li><strong>Initialize the output folder</strong>:</li> </ol> <p>This can be done using the following command:</p> <div class=highlight><pre><span></span><code>%<span class=w> </span>aspen<span class=w> </span>-m<span class=o>=</span>init<span class=w> </span>-w<span class=o>=</span>&lt;path_to_output_folder&gt;
</code></pre></div> <p>The above command will create <code>config.yaml</code> and <code>samples.tsv</code> in the output folder. Please edit these as per your requirements. You can replace the <code>samples.tsv</code> file in the output folder with the sample manifest created in the previous step outlined above. <code>contrasts.tsv</code> should also be included if running differential ATAC.</p> <ol> <li><strong>Dryrun</strong>:</li> </ol> <p>To dry-run the pipeline, you can run the following command after initializing the output folder:</p> <div class=highlight><pre><span></span><code>%<span class=w> </span>aspen<span class=w> </span>-m<span class=o>=</span>dryrun<span class=w> </span>-w<span class=o>=</span>&lt;path_to_output_folder&gt;
</code></pre></div> <p>This should list out the chain of jobs (DAG) that will be submitted to the job scheduler.</p> <ol> <li><strong>RUN!!</strong>:</li> </ol> <p>If the dry-run looks as expected, then you can submit the job using:</p> <div class=highlight><pre><span></span><code>%<span class=w> </span>aspen<span class=w> </span>-m<span class=o>=</span>run<span class=w> </span>-w<span class=o>=</span>&lt;path_to_output_folder&gt;
</code></pre></div> <p>This will submit one <em>master</em> job to slurm, which will in turn keep managing the entire pipeline and submit/monitor jobs to the job scheduler as and when required.</p> <h3 id=3-genomes>3. Genomes<a class=headerlink href=#3-genomes title="Permanent link">&para;</a></h3> <p>ASPEN supports the following genome versions:</p> <table> <thead> <tr> <th>Genome Version</th> <th>Organism</th> </tr> </thead> <tbody> <tr> <td>hg38</td> <td>Human</td> </tr> <tr> <td>hg19</td> <td>Human</td> </tr> <tr> <td>mm10</td> <td>Mouse</td> </tr> <tr> <td>mmul10</td> <td>Macaca mulatta(Rhesus monkey) or rheMac10</td> </tr> <tr> <td>bosTau9</td> <td>Bos taurus(cattle)</td> </tr> </tbody> </table> <h3 id=4-disclaimer>4. Disclaimer<a class=headerlink href=#4-disclaimer title="Permanent link">&para;</a></h3> <p>This snakemake pipeline is built to run on <a href=https://hpc.nih.gov/ >Biowulf</a> and <a href=https://ncifrederick.cancer.gov/staff/frce/ >FRCE</a>. But, as it uses containers for all intermediate steps, the pipeline can be executed on any HPC with minimal edits to the config file.</p> <h3 id=5-help>5. Help<a class=headerlink href=#5-help title="Permanent link">&para;</a></h3> <p>For comments/suggestions/advice please reach out to <a href=mailto:vishal.koparde@nih.gov>Vishal Koparde</a> or <a href=mailto:CCBR_Pipeliner@mail.nih.gov>CCBR_Pipeliner</a>. You can also open a new issue <a href=https://github.com/CCBR/ASPEN/issues>here</a>.</p> <hr> <p align=center> <a href=#aspen>Back to Top</a> </p> <hr> <!--- write additional info that you only want to appear in the docs index here --> <hr> <div class=md-source-file> <small> Last update: 2024-01-16 </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2023 CCBR </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=http://bioinformatics.cancer.gov target=_blank rel=noopener title=bioinformatics.cancer.gov class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M144 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zm368 0a80 80 0 1 1 0 160 80 80 0 1 1 0-160zM0 298.7C0 239.8 47.8 192 106.7 192h42.7c15.9 0 31 3.5 44.6 9.7-1.3 7.2-1.9 14.7-1.9 22.3 0 38.2 16.8 72.5 43.3 96H21.3C9.6 320 0 310.4 0 298.7zM405.3 320h-.7c26.6-23.5 43.3-57.8 43.3-96 0-7.6-.7-15-1.9-22.3 13.6-6.3 28.7-9.7 44.6-9.7h42.7c58.9 0 106.7 47.8 106.7 106.7 0 11.8-9.6 21.3-21.3 21.3H405.3zM224 224a96 96 0 1 1 192 0 96 96 0 1 1-192 0zm-96 261.3c0-73.6 59.7-133.3 133.3-133.3h117.4c73.6 0 133.3 59.7 133.3 133.3 0 14.7-11.9 26.7-26.7 26.7H154.7c-14.7 0-26.7-11.9-26.7-26.7z"/></svg> </a> <a href=https://github.com/CCBR target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/orgs/nciccbr/repositories target=_blank rel=noopener title=hub.docker.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "search": "assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script> <script src=assets/javascripts/bundle.51198bba.min.js></script> </body> </html>