<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content=CCBR><link rel="shortcut icon" href=../img/favicon.ico><title>Overview of ASPEN - ASPEN</title><link rel=stylesheet href=../css/theme.css><link rel=stylesheet href=../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../css/custom.css rel=stylesheet><link href=../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Overview of ASPEN";
        var mkdocs_page_input_path = "overview.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=.. class="icon icon-home"> ASPEN </a><div role=search> <form id=rtd-search-form class=wy-form action=../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=..>Background</a> </li> </ul> <ul> <li class=toctree-l1><a href=../introduction/ class="reference internal">Introduction to ATAC-seq</a> </li> </ul> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Overview of ASPEN</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#data-preprocessing>Data Preprocessing</a> <ul> <li class=toctree-l3><a class="reference internal" href=#adapter-trimming>Adapter Trimming</a> </li> <li class=toctree-l3><a class="reference internal" href=#quality-assessment>Quality Assessment</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#alignment-and-filtering>Alignment and Filtering</a> <ul> <li class=toctree-l3><a class="reference internal" href=#blacklist-filtering>Blacklist Filtering</a> </li> <li class=toctree-l3><a class="reference internal" href=#read-alignment>Read Alignment</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#peak-calling-and-annotation>Peak Calling and Annotation</a> <ul> <li class=toctree-l3><a class="reference internal" href=#peak-detection>Peak Detection</a> </li> <li class=toctree-l3><a class="reference internal" href=#consensus-peaks>Consensus Peaks</a> </li> <li class=toctree-l3><a class="reference internal" href=#peak-annotation>Peak Annotation</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#quality-control-metrics>Quality Control Metrics</a> <ul> <li class=toctree-l3><a class="reference internal" href=#fragment-length-distribution>Fragment Length Distribution</a> </li> <li class=toctree-l3><a class="reference internal" href=#library-complexity>Library Complexity</a> </li> <li class=toctree-l3><a class="reference internal" href=#transcription-start-site-tss-enrichment>Transcription Start Site (TSS) Enrichment</a> </li> <li class=toctree-l3><a class="reference internal" href=#fraction-of-reads-in-peaks-frip>Fraction of Reads in Peaks (FRiP)</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#motif-enrichment-analysis>Motif Enrichment Analysis</a> <ul> <li class=toctree-l3><a class="reference internal" href=#homer-and-ame>HOMER and AME</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#comprehensive-reporting>Comprehensive Reporting</a> <ul> <li class=toctree-l3><a class="reference internal" href=#multiqc-integration>MultiQC Integration</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#differential-accessiblity-analyses>Differential Accessiblity Analyses</a> <ul> <li class=toctree-l3><a class="reference internal" href=#refining-peak-calls-to-fixed-width-peaks>Refining Peak Calls to Fixed-Width Peaks</a> </li> <li class=toctree-l3><a class="reference internal" href=#defining-regions-of-interest-rois>Defining Regions of Interest (ROIs)</a> </li> <li class=toctree-l3><a class="reference internal" href=#quantification-of-differential-chromatin-accessibility>Quantification of Differential Chromatin Accessibility</a> <ul> <li class=toctree-l4><a class="reference internal" href=#tn5-nicking-sites-count-matrix>Tn5 Nicking Sites Count Matrix</a> </li> <li class=toctree-l4><a class="reference internal" href=#differential-accessibility-analysis-with-deseq2>Differential Accessibility Analysis with DESeq2</a> </li> <li class=toctree-l4><a class="reference internal" href=#integration-with-gene-annotations>Integration with Gene Annotations</a> </li> </ul> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#reporting>Reporting</a> </li> </ul> </li> </ul> <ul> <li class=toctree-l1><a href=../deployment/ class="reference internal">Running ASPEN</a> </li> </ul> <ul> <li class=toctree-l1><a href=../outputs/ class="reference internal">ASPEN Output</a> </li> </ul> <ul> <li class=toctree-l1><a href=../limitations/ class="reference internal">Limitations of ASPEN</a> </li> </ul> <ul> <li class=toctree-l1><a href=../communications/ class="reference internal">Communications</a> </li> </ul> <ul> <li class=toctree-l1><a href=../changelog/ class="reference internal">Change Log</a> </li> </ul> <ul> <li class=toctree-l1><a href=../contributing/ class="reference internal">Contributing to ASPEN</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=..>ASPEN</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=.. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Overview of ASPEN</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/ASPEN/edit/main/docs/overview.md>Edit on CCBR/ASPEN</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=overview-of-aspen>Overview of ASPEN<a class=headerlink href=#overview-of-aspen title="Permanent link">&para;</a></h1> <p>To address <a href=../introduction/#challenges-in-atac-seq-data-analysis>these</a> challenges, CCBR developed ASPEN (<strong>A</strong>tac <strong>S</strong>eq <strong>P</strong>ip<strong>E</strong>li<strong>N</strong>e), an automated pipeline tailored for the comprehensive analysis of ATAC-seq data. ASPEN is designed to process paired-end Illumina sequencing data, guiding users from raw data through quality control, alignment, peak calling, and downstream analyses. By integrating a suite of established bioinformatics tools and adhering to best practices, ASPEN ensures robust, reproducible, and high-quality results. For more information on the challenges in ATAC-seq data analysis, refer to the [Challenges in ATAC-seq Data Analysis] section.</p> <h2 id=data-preprocessing><strong>Data Preprocessing</strong><a class=headerlink href=#data-preprocessing title="Permanent link">&para;</a></h2> <h3 id=adapter-trimming><strong>Adapter Trimming</strong><a class=headerlink href=#adapter-trimming title="Permanent link">&para;</a></h3> <p>Utilizes CutAdapt to remove adapter sequences from raw reads, ensuring that subsequent analyses are not confounded by extraneous sequences.</p> <h3 id=quality-assessment><strong>Quality Assessment</strong><a class=headerlink href=#quality-assessment title="Permanent link">&para;</a></h3> <p>Employs FastQC for initial quality checks and MultiQC for aggregated reporting, providing comprehensive insights into data quality and highlighting potential issues such as low-quality bases or GC content biases.</p> <h2 id=alignment-and-filtering><strong>Alignment and Filtering</strong><a class=headerlink href=#alignment-and-filtering title="Permanent link">&para;</a></h2> <h3 id=blacklist-filtering><strong>Blacklist Filtering</strong><a class=headerlink href=#blacklist-filtering title="Permanent link">&para;</a></h3> <p>Removes reads aligning to known blacklisted regions and mitochondria, reducing potential biases in peak detection and ensuring that analyses focus on biologically relevant regions.</p> <h3 id=read-alignment><strong>Read Alignment</strong><a class=headerlink href=#read-alignment title="Permanent link">&para;</a></h3> <p>Aligns reads to the reference genome using Bowtie2, generating multiple output files, including <code>qsorted.bam</code>/<code>tagAlign.gz</code> for peak calling and <code>dedup.bam</code> for downstream analyses.</p> <h2 id=peak-calling-and-annotation><strong>Peak Calling and Annotation</strong><a class=headerlink href=#peak-calling-and-annotation title="Permanent link">&para;</a></h2> <h3 id=peak-detection><strong>Peak Detection</strong><a class=headerlink href=#peak-detection title="Permanent link">&para;</a></h3> <p>Implements MACS2 and Genrich to identify regions of open chromatin, accommodating both narrow and broad peak detection to capture a wide range of regulatory elements. Default MACS2 are selected from ENCODE. Peak calling uses lenient q-value setting and a q-value filter is then applied on the <code>.narrowPeak</code> file.</p> <h3 id=consensus-peaks><strong>Consensus Peaks</strong><a class=headerlink href=#consensus-peaks title="Permanent link">&para;</a></h3> <p>For datasets with multiple replicates, ASPEN generates consensus peak sets based on simple overlaps between replicates and keeping peaks represented in &gt;= 50% of replicates. Thus enhancing the reliability and reproducibility of detected regions by focusing on peaks consistently observed across replicates.</p> <h3 id=peak-annotation><strong>Peak Annotation</strong><a class=headerlink href=#peak-annotation title="Permanent link">&para;</a></h3> <p>Utilizes ChIPseeker to annotate peaks, providing context regarding their genomic locations and potential regulatory roles, such as associations with promoters, enhancers, or gene bodies.</p> <h2 id=quality-control-metrics><strong>Quality Control Metrics</strong><a class=headerlink href=#quality-control-metrics title="Permanent link">&para;</a></h2> <h3 id=fragment-length-distribution><strong>Fragment Length Distribution</strong><a class=headerlink href=#fragment-length-distribution title="Permanent link">&para;</a></h3> <p>Custom scripts are utilized to analyze insert size to assess nucleosome positioning and sample quality, with characteristic patterns indicating the presence of nucleosome-free regions and mono- or di-nucleosomes.</p> <h3 id=library-complexity><strong>Library Complexity</strong><a class=headerlink href=#library-complexity title="Permanent link">&para;</a></h3> <p>Estimates complexity using Preseq, aiding in the evaluation of sequencing depth sufficiency and potential PCR biases.</p> <h3 id=transcription-start-site-tss-enrichment><strong>Transcription Start Site (TSS) Enrichment</strong><a class=headerlink href=#transcription-start-site-tss-enrichment title="Permanent link">&para;</a></h3> <p>Calculates TSS enrichment scores, serving as indicators of data quality by measuring the accumulation of reads around transcription start sites, a hallmark of open chromatin.</p> <h3 id=fraction-of-reads-in-peaks-frip><strong>Fraction of Reads in Peaks (FRiP)</strong><a class=headerlink href=#fraction-of-reads-in-peaks-frip title="Permanent link">&para;</a></h3> <p>Determines the proportion of reads within called peaks, reflecting the signal-to-noise ratio and overall quality of the dataset. Fractions if reads in Promoters, Enhancers and DHS (DNase Hyper Sensitivity) regions are also computed.</p> <h2 id=motif-enrichment-analysis><strong>Motif Enrichment Analysis</strong><a class=headerlink href=#motif-enrichment-analysis title="Permanent link">&para;</a></h2> <h3 id=homer-and-ame><strong>HOMER and AME</strong><a class=headerlink href=#homer-and-ame title="Permanent link">&para;</a></h3> <p>Conducts motif enrichment analyses to identify potential transcription factor binding sites within accessible regions, offering insights into regulatory mechanisms and facilitating the discovery of key drivers of gene expression.</p> <h2 id=comprehensive-reporting><strong>Comprehensive Reporting</strong><a class=headerlink href=#comprehensive-reporting title="Permanent link">&para;</a></h2> <h3 id=multiqc-integration><strong>MultiQC Integration</strong><a class=headerlink href=#multiqc-integration title="Permanent link">&para;</a></h3> <p>Generates an extensive HTML report consolidating all QC metrics and analysis results, facilitating easy interpretation, visualization, and sharing of findings.</p> <h2 id=differential-accessiblity-analyses><strong>Differential Accessiblity Analyses</strong><a class=headerlink href=#differential-accessiblity-analyses title="Permanent link">&para;</a></h2> <p>In addition to its core functionalities, ASPEN offers advanced capabilities for differential accessibility analysis, enabling researchers to identify and interpret regions of the genome that exhibit significant changes in chromatin accessibility under different conditions. This process involves several key steps: refining peak calls to fixed-width peaks, defining regions of interest (ROIs), quantifying Tn5 transposase insertion event counts, performing statistical analysis to detect differential signals, and integrating these findings with ChIP-seq annotations for comprehensive biological insights.</p> <h3 id=refining-peak-calls-to-fixed-width-peaks><strong>Refining Peak Calls to Fixed-Width Peaks</strong><a class=headerlink href=#refining-peak-calls-to-fixed-width-peaks title="Permanent link">&para;</a></h3> <p>After the initial peak calling, ASPEN refines these peaks by centering them on their summits and extending them to a uniform width of 500 base pairs (bp). This standardization facilitates consistent downstream analyses and comparisons across different datasets. The approach of extending peak summits by Â±250 bp to achieve a fixed width of 500 bp is a common practice in ATAC-seq data analysis, as it focuses on the most accessible regions of the chromatin and reduces variability in peak sizes. </p> <h3 id=defining-regions-of-interest-rois><strong>Defining Regions of Interest (ROIs)</strong><a class=headerlink href=#defining-regions-of-interest-rois title="Permanent link">&para;</a></h3> <p>ASPEN processes fixed width peak calls on each replicate across all samples. Subsequently, it generates a union of all identified peaks to create comprehensive regions of interest (ROIs). The pipeline produces <code>ROI.bed</code> and <code>ROI.gtf</code> files for each of the peak callers, which are utilized for quantifying counts using featureCounts. This approach ensures a consistent framework for downstream differential accessibility analysis, enhancing the robustness and interpretability of the results.</p> <h3 id=quantification-of-differential-chromatin-accessibility><strong>Quantification of Differential Chromatin Accessibility</strong><a class=headerlink href=#quantification-of-differential-chromatin-accessibility title="Permanent link">&para;</a></h3> <p>Once the ROIs are established, ASPEN generates Tn5 nicking sites count matrices:</p> <h4 id=tn5-nicking-sites-count-matrix><strong>Tn5 Nicking Sites Count Matrix</strong><a class=headerlink href=#tn5-nicking-sites-count-matrix title="Permanent link">&para;</a></h4> <p>This matrix quantifies the frequency of Tn5 transposase insertion events at each ROI. Reads used by MACS2 and reads coming out of Genrich are used to compute Tn5 sites and counting them. The Tn5 transposase preferentially inserts into accessible regions of the chromatin, and the number of insertion events serves as a proxy for chromatin accessibility. By counting these insertion sites, researchers can accurately infer the openness of chromatin regions under different experimental conditions.</p> <h4 id=differential-accessibility-analysis-with-deseq2><strong>Differential Accessibility Analysis with DESeq2</strong><a class=headerlink href=#differential-accessibility-analysis-with-deseq2 title="Permanent link">&para;</a></h4> <p>To identify statistically significant differences in chromatin accessibility between conditions, ASPEN employs DESeq2, a widely-used tool for differential analysis of count data. DESeq2 models the count data to detect changes in accessibility, providing robust statistical inferences even with complex experimental designs. By comparing the count matrices across conditions, DESeq2 identifies ROIs with significant differential accessibility, shedding light on regulatory elements that may be functionally relevant to the experimental context.</p> <h4 id=integration-with-gene-annotations><strong>Integration with Gene Annotations</strong><a class=headerlink href=#integration-with-gene-annotations title="Permanent link">&para;</a></h4> <p>To enhance the biological interpretation of differential accessibility results, ASPEN integrates the findings with peak and gene annotations from ChIPSeeker, offering insights into the regulatory landscape of the genome. This integration allows researchers to identify genes located near differentially accessible peaks, which can be crucial for understanding the functional implications of chromatin accessibility changes. By associating these peaks with nearby genes, scientists can infer potential regulatory relationships and gain a deeper understanding of the underlying molecular mechanisms.</p> <h2 id=reporting><strong>Reporting</strong><a class=headerlink href=#reporting title="Permanent link">&para;</a></h2> <p>ASPEN enhances differential chromatin accessibility analysis by providing an interactive HTML report generated from DESeq2 results, featuring various visualizations. Additionally, it offers a TSV (tabl-delimited) file with integrated gene annotations, compatible with Microsoft Excel, enabling efficient data manipulation and facilitating the identification of genes near regions with altered accessibility. The Excel file is aggregated accross all different contrasts queried in the project.</p> </div> </div><footer> <div class=rst-footer-buttons role=navigation aria-label="Footer Navigation"> <a href=../introduction/ class="btn btn-neutral float-left" title="Introduction to ATAC-seq"><span class="icon icon-circle-arrow-left"></span> Previous</a> <a href=../deployment/ class="btn btn-neutral float-right" title="Running ASPEN">Next <span class="icon icon-circle-arrow-right"></span></a> </div> <hr> <div role=contentinfo> <!-- Copyright etc --> <p>Copyright &copy; 2023 CCBR</p> </div> Built with <a href=https://www.mkdocs.org/ >MkDocs</a> using a <a href=https://github.com/readthedocs/sphinx_rtd_theme>theme</a> provided by <a href=https://readthedocs.org>Read the Docs</a>. </footer> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/ASPEN class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/ASPEN</a> </span> <span><a href=../introduction/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../deployment/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../js/jquery-3.6.0.min.js></script> <script>var base_url = "..";</script> <script src=../js/theme_extra.js></script> <script src=../js/theme.js></script> <script src=../js/custom.js></script> <script src=../search/main.js></script> <script src=../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 