# ðŸš€ ASPEN Outputs

## ðŸ“‚ Workdir

The workdir which is supplied as `-w` while running aspen `init`, `dryrun` and `run` commands will contain the following files:

```bash
WORKDIR
â”œâ”€â”€ cluster.json
â”œâ”€â”€ config.yaml
â”œâ”€â”€ contrasts.tsv
â”œâ”€â”€ dryrun_git_commit.txt
â”œâ”€â”€ dryrun.log
â”œâ”€â”€ fastqs
â”œâ”€â”€ logs
â”œâ”€â”€ results
â”œâ”€â”€ run_git_commit.txt
â”œâ”€â”€ runinfo.yaml
â”œâ”€â”€ runslurm_snakemake_report.html
â”œâ”€â”€ sampleinfo.txt
â”œâ”€â”€ samples.tsv
â”œâ”€â”€ scripts
â”œâ”€â”€ slurm-XXXXXXX.out
â”œâ”€â”€ snakemake.log
â”œâ”€â”€ snakemake.log.jobby
â”œâ”€â”€ snakemake.log.jobby.short
â”œâ”€â”€ snakemake.stats
â”œâ”€â”€ submit_script.sbatch
â””â”€â”€ tools.yaml
```

Here are more details about these files:

| **File**                         | **File Type** | **Mode (`-m`) When This File is Created/Overwritten** | **Description**                                                                                                                                                       |
| -------------------------------- | ------------- | ----------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `cluster.json`                   | JSON          | init                                                  | Defines cluster resources per snakemake rule; this file can be edited to override default computate resource allocations per snakemake rule                           |
| `config.yaml`                    | YAML          | init; can be edited later                             | Configurable parameters for this specific run                                                                                                                         |
| `contrasts.tsv`                  | TSV           | Needs to be added in after init                       | List of contrasts to run, one per line; has no header                                                                                                                 |
| `dryrun_git_commit.txt`          | TXT           | dryrun                                                | The git commit hash of the version of ASPEN used at dryrun                                                                                                            |
| `dryrun.log`                     | TXT           | dryrun                                                | Log from `-m=dryrun`                                                                                                                                                  |
| `fastqs`                         | FOLDER        | dryrun                                                | Folder containing symlinks to raw data                                                                                                                                |
| `logs`                           | FOLDER        | dryrun                                                | Folder containing all logs including Slurm `.out` and `.err` files. Also contains older timestamped `runinfo.yaml` and `snakemake.stats` files.                       |
| `results`                        | FOLDER        | Created at dryrun but populated during run            | Main outputs folder                                                                                                                                                   |
| `runinfo.yaml`                   | YAML          | After completion of run                               | Metadata about the run executor, etc.                                                                                                                                 |
| `runslurm_snakemake_report.html` | HTML          | After completion of run                               | HTML report including DAG and resource utilization                                                                                                                    |
| `sampleinfo.txt`                 | TXT           | dryrun, run                                           | Tab-delimited mappings between `replicateNames` and `sampleNames`                                                                                                     |
| `samples.tsv`                    | TSV           | init; can be edited later                             | Tab-delimited manifest with `replicateName`, `sampleName`, `path_to_R1_fastq`, `path_to_R2_fastq`. This file has a header.                                            |
| `scripts`                        | FOLDER        | init                                                  | Folder keeps local copy of scripts called by various rules                                                                                                            |
| `run_git_commit.txt`             | TXT           | run                                                   | The git commit hash of the version of ASPEN used at run                                                                                                               |
| `slurm-XXXXXXX.out`              | TXT           | run                                                   | Slurm `.out` file for the master job                                                                                                                                  |
| `snakemake.log`                  | TXT           | run                                                   | Snakemake `.log` file for the master job; older copies timestamped and moved into `logs` folder                                                                       |
| `snakemake.stats`                | JSON          | run                                                   | per rule runtime stats                                                                                                                                                |
| `submit_script.sbatch`           | TXT           | run                                                   | Slurm script to kickstart the main Snakemake job                                                                                                                      |
| `tools.yaml`                     | YAML          | run                                                   | YAML containing the version of tools used in the pipeline (obsolete; was used to load specific module versions prior to moving over to Docker/Singularity containers) |

## ðŸ“Š `results` folder

The results directory contains the actual output files. Below are the folders that you may find within it.

```bash
WORKDIR
â”œâ”€â”€ results
    â”œâ”€â”€ alignment
    â”‚Â Â  â”œâ”€â”€ dedupBam
    â”‚Â Â  â”œâ”€â”€ filteredBam
    â”‚Â Â  â”œâ”€â”€ qsortedBam
    â”‚Â Â  â””â”€â”€ tagAlign
    â”œâ”€â”€ peaks
    â”‚Â Â  â”œâ”€â”€ genrich
    â”‚Â Â  â”‚Â Â  â”œâ”€â”€ DiffATAC
    â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reads
    â”‚Â Â  â”‚Â Â  â”‚Â Â  â””â”€â”€ tn5sites
    â”‚Â Â  â”‚Â Â  â””â”€â”€ fixed_width
    â”‚Â Â  â””â”€â”€ macs2
    â”‚Â Â      â”œâ”€â”€ DiffATAC
    â”‚Â Â      â”‚Â Â  â”œâ”€â”€ reads
    â”‚Â Â      â”‚Â Â  â””â”€â”€ tn5sites
    â”‚Â Â      â””â”€â”€ fixed_width
    â”œâ”€â”€ QC
    â”‚Â Â  â”œâ”€â”€ fastqc
    â”‚Â Â  â”œâ”€â”€ fld
    â”‚Â Â  â”œâ”€â”€ FQscreen
    â”‚Â Â  â”œâ”€â”€ frip
    â”‚Â Â  â”œâ”€â”€ multiqc_data
    â”‚Â Â  â”œâ”€â”€ peak_annotation
    â”‚Â Â  â”œâ”€â”€ preseq
    â”‚Â Â  â””â”€â”€ tss
    â”œâ”€â”€ spikein
    â”‚Â Â  â”œâ”€â”€ <sample_1>
    â”‚Â Â  â”œâ”€â”€ <sample_2>
    â”‚Â Â  â”œâ”€â”€ <sample_3>
    â”‚   â”‚ ...
    â”‚Â Â  â””â”€â”€ <sample_n>
    â”œâ”€â”€ tmp
    â”‚Â Â  â”œâ”€â”€ BL
    â”‚Â Â  â”œâ”€â”€ genrichReads
    â”‚Â Â  â””â”€â”€ trim
    â””â”€â”€ visualization
        â”œâ”€â”€ reads_bam
        â”œâ”€â”€ reads_bed
        â”œâ”€â”€ reads_bigwig
        â”œâ”€â”€ tn5sites_bam
        â””â”€â”€ tn5sites_bigwig
```

Content details:

| Folder        | SubFolder           | Description                                                                                                                                                                                                                                         |
| ------------- | ------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| alignment     | qsortedBam          | - Query sorted Bowtie2 alignments in BAM format. <br> - Excludes unmapped and platform/vendor quality failing reads. <br> - Used for Genrich peak calling.                                                                                          |
| alignment     | filteredBam         | - Filtered BAM files after excluding non-primary, supplementary, and MAPQ <=5 alignments. <br> - Used for counting reads/tn5 nicks. <br> - Derived from `qsortedBam`.                                                                               |
| alignment     | dedupBam            | - Deduplicated filtered BAM files. <br> - PCR or optical duplicates marked with PicardTools and excluded. <br> - Can be used downstream with CCBR_TOBIAS pipeline. <br> - Derived from `filteredBam`.                                               |
| alignment     | tagAlign            | - `tagAlign.gz` files used for MACS2 peak calling. <br> - Derived from `dedupBam`.                                                                                                                                                                  |
| peaks         | genrich & macs      | - Genrich/MACS2 peak calls (raw, consensus, fixed-width). <br> - Contains ROI files with Diff-ATAC results if `contrasts.tsv` is provided. <br> - Calculated with DESeq2 using both read counts and tn5 nicking sites in ROI.                       |
| QC            | various             | - Flagstats. <br> - Dupmetrics. <br> - Read counts. <br> - Motif enrichments. <br> - FLD stats. <br> - Fqscreen. <br> - FRiP. <br> - ChIPSeeker results. <br> - TSS enrichments. <br> - Preseq. <br> - Homer/AME motif enrichments. <br> - MultiQC. |
| QC            | peak_annotation     | detailed peak annotations described below                                                                                                                                                                                                           |
| spikein       | 1 folder per sample | - Per sample spike-in counts. <br> - Overall scaling factors table.                                                                                                                                                                                 |
| tmp           | various             | - Can be deleted. <br> - Blacklist index. <br> - Intermediate FASTQs. <br> - Genrich output reads.                                                                                                                                                  |
| visualization | reads_bam           | - Tn5 nick adjusted reads in BAM format. <br> - Derived from `filteredBam`.                                                                                                                                                                         |
| visualization | reads_bed           | - Tn5 nick adjusted reads in BED format. <br> - Derived from `reads_bam`. <br> - Can be used by ChromVar.                                                                                                                                           |
| visualization | reads_bigwig        | - Tn5 nick adjusted reads in BIGWIG format. <br> - Scaled using spike-in scaling factors if present. <br> - Derived from `reads_bam`.                                                                                                               |
| visualization | tn5sites_bam        | - Tn5 nicking sites in BAM format. <br> - Derived from `filteredBam`.                                                                                                                                                                               |
| visualization | tn5sites_bigwig     | - Tn5 nicking sites in BIGWIG format. <br> - Scaled using spike-in scaling factors if present. <br> - Derived from `tn5sites_bam`.                                                                                                                  |

!!! note
BAM files from `dedupBam` can be used for downstream footprinting analysis using [CCBR_TOBIAS](https://github.com/CCBR/CCBR_Tobias) pipeline

!!! note
[bamCompare](https://deeptools.readthedocs.io/en/develop/content/tools/bamCompare.html) from deeptools can be run to compare BAMs from `dedupBam` for comprehensive BAM comparisons.

!!! note
BAM files from `dedupBam` can also be converted to BED format and processed with [chromVAR](https://github.com/GreenleafLab/chromVAR) to identify variability in motif accessibility across samples and assess differentially active transcription factors from the JASPAR database.

#### Peak Annotation folder

This folder will contain ChIPseeker results for:

- individual replicate `*.narrowPeak` files
- `*.consensus.bed` files
- `*.fixed_width.consensus.narrowPeak` files

The `QC` folder contains the `multiqc_report.html` file which provides a comprehensive summary of the quality control metrics across all samples, including read quality, duplication rates, and other relevant statistics. This report aggregates results from various QC tools such as FastQC, FastqScreen, FLD, TSS enrichment, Peak Annotations, and others, presenting them in an easy-to-read format with interactive plots and tables. It helps in quickly identifying any issues with the sequencing data and ensures that the data quality is sufficient for downstream analysis.

| File                                  | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `*.narrowPeak.annotated.gz`           | peak calls annotated using ChIPseeker, gzipped                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `*.narrowPeak.annotated.distribution` | annotation bins : <br> - **3'UTR**: No. of peaks in the 3' untranslated region. <br> - **5'UTR**: No. of peaks in the 5' untranslated region. <br> - **Distal Intergenic**: No. of peaks in distal intergenic regions. <br> - **Downstream (<1kb)**: No. of peaks annotated downstream within 1kb. <br> - **Downstream (1-2kb)**: No. of peaks annotated downstream between 1-2kb. <br> - **Downstream (2-3kb)**: No. of peaks annotated downstream between 2-3kb. <br> - **Promoter (<=1kb)**: No. of peaks in promoters within 1kb. <br> - **Promoter (1-2kb)**: No. of peaks in promoters between 1-2kb. <br> - **Exon**: No. of peaks in exonic regions. |
| `*.narrowPeak.annotated_summary`      | More stats on each of the above bins .. like: <br> - medianWidth <br> - medianpValue <br> - medianqValue                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `*.narrowPeak.genelist`               | ensemblID and gene symbols of genes with peaks in their promoter regions (including 5' UTR)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |

### MACS2 output folder

For a typical 2 sample analysis with 2 replicates each this folder should look like this:

```bash
WORKDIR
â”œâ”€â”€ results
    â”œâ”€â”€ peaks
        â””â”€â”€ macs2
            â”œâ”€â”€ sample1
            â”‚Â Â  â”œâ”€â”€ sample1_replicate1.macs2.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample1_replicate1.macs2.narrowPeak_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample1_replicate1.macs2.unfiltered.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample1_replicate2.macs2.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample1_replicate2.macs2.narrowPeak_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample1_replicate2.macs2.unfiltered.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample1.macs2.consensus.bed
            â”‚Â Â  â”œâ”€â”€ sample1.macs2.consensus.bed_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample1.macs2.pooled.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample1.macs2.pooled_summits.bed
            â”‚Â Â  â””â”€â”€ sample1.macs2.pooled.unfiltered.narrowPeak
            â”œâ”€â”€ sample1.consensus.macs2.peakfiles
            â”œâ”€â”€ sample1.replicate.macs2.peakfiles
            â”œâ”€â”€ DiffATAC
            â”‚Â Â  â”œâ”€â”€ reads
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ all_diff_atacs.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ all_diff_atacs.tsv
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ degs.done
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ sample2_vs_sample1.html
            â”‚Â Â  â”‚Â Â  â””â”€â”€ sample2_vs_sample1.tsv
            â”‚Â Â  â””â”€â”€ tn5sites
            â”‚Â Â      â”œâ”€â”€ all_diff_atacs.html
            â”‚Â Â      â”œâ”€â”€ all_diff_atacs.tsv
            â”‚Â Â      â”œâ”€â”€ degs.done
            â”‚Â Â      â”œâ”€â”€ sample2_vs_sample1.html
            â”‚Â Â      â””â”€â”€ sample2_vs_sample1.tsv
            â”œâ”€â”€ sample2
            â”‚Â Â  â”œâ”€â”€ sample2_replicate1.macs2.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample2_replicate1.macs2.narrowPeak_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample2_replicate1.macs2.unfiltered.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample2_replicate2.macs2.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample2_replicate2.macs2.narrowPeak_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample2_replicate2.macs2.unfiltered.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample2.macs2.consensus.bed
            â”‚Â Â  â”œâ”€â”€ sample2.macs2.consensus.bed_motif_enrichment
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ame_results.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ background.fa
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.html
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ knownResults.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ motifFindingParameters.txt
            â”‚Â Â  â”‚Â Â  â”œâ”€â”€ seq.autonorm.tsv
            â”‚Â Â  â”‚Â Â  â””â”€â”€ target.fa
            â”‚Â Â  â”œâ”€â”€ sample2.macs2.pooled.narrowPeak
            â”‚Â Â  â”œâ”€â”€ sample2.macs2.pooled_summits.bed
            â”‚Â Â  â””â”€â”€ sample2.macs2.pooled.unfiltered.narrowPeak
            â”œâ”€â”€ sample2.consensus.macs2.peakfiles
            â”œâ”€â”€ sample2.replicate.macs2.peakfiles
            â””â”€â”€ fixed_width
                â”œâ”€â”€ sample1_replicate1.macs2.fixed_width.narrowPeak
                â”œâ”€â”€ sample1_replicate2.macs2.fixed_width.narrowPeak
                â”œâ”€â”€ sample1.fixed_width.consensus.narrowPeak
                â”œâ”€â”€ sample1.renormalized.fixed_width.consensus.narrowPeak
                â”œâ”€â”€ sample1.renormalized.fixed_width.consensus.narrowPeak.annotated.gz
                â”œâ”€â”€ counts
                â”‚Â Â  â”œâ”€â”€ ROI.macs2.reads_counts.tsv
                â”‚Â Â  â””â”€â”€ ROI.macs2.tn5sites_counts.tsv
                â”œâ”€â”€ sample2_replicate1.macs2.fixed_width.narrowPeak
                â”œâ”€â”€ sample2_replicate2.macs2.fixed_width.narrowPeak
                â”œâ”€â”€ sample2.fixed_width.consensus.narrowPeak
                â”œâ”€â”€ sample2.renormalized.fixed_width.consensus.narrowPeak
                â”œâ”€â”€ sample2.renormalized.fixed_width.consensus.narrowPeak.annotated.gz
                â”œâ”€â”€ ROI.macs2.bed
                â”œâ”€â”€ ROI.macs2.bed.annotated.gz
                â”œâ”€â”€ ROI.macs2.bed.annotated.gz.gz
                â”œâ”€â”€ ROI.macs2.bed.annotation_distribution
                â”œâ”€â”€ ROI.macs2.bed.annotation_summary
                â”œâ”€â”€ ROI.macs2.bed.genelist
                â”œâ”€â”€ ROI.macs2.gtf
                â”œâ”€â”€ ROI.macs2.narrowPeak
                â”œâ”€â”€ ROI.macs2.renormalized.narrowPeak
                â””â”€â”€ Rplots.pdf
```

Some of the key output files are:

| File                                                | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `*.macs2.narrowPeak`                                | peak calls from MACS2 filtered by q-value for each samples each replicate                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| `*.macs2.unfiltered.narrowPeak`                     | peak calls from MACS2 (unfiltered) for each samples each replicate                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| `*.narrowPeak_motif_enrichment/ame_results.txt`     | motif enrichment results from AME tool from MEME suite using HOCOMOCO v11 database                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| `*.narrowPeak_motif_enrichment/knownResults.txt`    | motif enrichment results using HOMER with HOCOMOCO v11 database                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| `*.macs2.consensus.bed`                             | consensus peak call between multiple replicates of each sample. **Note:** consensus bed annotations are located in `QC/peak_annotations`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `DiffATAC/reads`                                    | folder containing differential open chromatin results: <br> - computated using read counts in MACS2 regions of interest (ROIs) <br> - `all_diff_atacs.html` HTML report aggregated across all contrasts from `contrasts.tsv` <br> - `all_diff_atacs.tsv` DESeq2 results in TSV format aggregated across all contrasts from `contrasts.tsv` <br> - HTML and TSV file each per contrast in `contrasts.tsv`                                                                                                                                                                                                                                   |
| `DiffATAC/tn5sites`                                 | folder containing differential open chromatin results: <br> - computated using Tn5 nicking site counts in MACS2 regions of interest (ROIs) <br> - `all_diff_atacs.html` HTML report aggregated across all contrasts from `contrasts.tsv` <br> - `all_diff_atacs.tsv` DESeq2 results in TSV format aggregated across all contrasts from `contrasts.tsv` <br> - HTML and TSV file each per contrast in `contrasts.tsv`                                                                                                                                                                                                                       |
| `fixed_width`                                       | `fixed_width` can be set in `config.yaml` to create peaks of a user defined fixed width (default 500bp). This folder contains: <br> - individual replicate `*.fixed_width.narrowPeak` files <br> - `*.renormalized.fixed_width.consensus.narrowPeak` per sample; [_Corces et. al._](https://doi.org/10.1038/nmeth.4396) method is used for consensus calling; used to generate MACS2 regions of interest (ROI) peaks which are used to generate a reads or Tn5 sites counts matrix for DESeq2 <br> - ROI related files: `ROI.macs2.bed`, `ROI.macs2.bed.annotated.gz`, `ROI.macs2.annotation_summary`, `ROI.macs2.annotation_distribution` |
| `fixed_width/counts/ROI.macs2.read_counts.tsv`      | read counts in MACS2 ROIs using featureCounts                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| `fixed_width/counts/ROI.reads_scaled_counts.tsv`    | `ROI.macs2.read_counts.tsv` scaled using spike-in scaling factors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| `fixed_width/counts/ROI.tn5sites_counts.tsv`        | Tn5 nicking site counts in MACS2 ROIs using featureCounts                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| `fixed_width/counts/ROI.tn5sites_scaled_counts.tsv` | `ROI.macs2.tn5sites_counts.tsv` scaled using spike-in scaling factors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |

### Genrich output folder

For a typical 2 sample analysis with 2 replicates each this folder should look like very similar to the MACS2 output structure described above.

## `logs` folder

This directory contains all .err and .out log files generated by SLURM for jobs submitted via Snakemake. Each file follows a consistent naming convention:

```bash
<SLURM_JOB_ID of master/head job>.<SLURM_JOB_ID of child job>.<Snakemake Rule Name>.<wildcard1_name=wildcard1_value,wildcard2_name=wildcard2_value>.<out or err>
```

This structure is particularly useful for troubleshooting and debugging, especially when the SLURM job IDs of failed jobs are known. By examining the corresponding .err or .out files, users can efficiently identify the source of errors within specific Snakemake rules and wildcards.

> DISCLAIMER: This folder hierarchy is significantly different than v1.0.6 and is subject to change with subsequent versions.
